### PR TITLE
karmada operator: support install karmada on remote cluster

### PR DESCRIPTION
**Pull request by calvin0327**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
support install karmada on remote cluster

**Which issue(s) this PR fixes**:
Fixes https://github.com/karmada-io/karmada/issues/3783

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`karmada operator`: Now able to install karmada on a remote cluster.
```

### CODE REVIEW

**Comment by calvin0327** [comment]
Test results:

* Create a karmada with a secret restore kubeconfig of remote cluster:

image

* Create the secret specified above:

image

* We login the remote cluster and check whether the karmada is be installed:

image

**Comment by calvin0327** [comment]
@RainbowMango @jwcesign I had test for this. PTAL

**Review by RainbowMango — COMMENTED** [review]
/assign
@calvin0327 Can you help resolve the conflict?

**Comment by calvin0327** [comment]
> /assign @calvin0327 Can you help resolve the conflict?

Ok, Done

File: operator/pkg/init.go [hunk 1]
```diff
@@ -36,6 +37,37 @@ type InitOptions struct {
...
 }
 
+// Validate is used to validate the initOptions before creating initJob.
+func (opt *InitOptions) Validate() error {
+	var errs []error
+
+	if len(opt.Name) == 0 || len(opt.Namespace) == 0 {
+		return errors.New("unexpected empty name or namespace")
+	}
+	if len(opt.CrdRemoteURL) > 0 {
```
Comment (jwcesign) [inline operator/pkg/init.go:47]: CRDRemoteURL

File: operator/pkg/init.go [hunk 2]
```diff
@@ -36,6 +37,37 @@ type InitOptions struct {
...
+		if _, err := url.Parse(opt.CrdRemoteURL); err != nil {
+			return fmt.Errorf("unexpected invalid crds remote url %s", opt.CrdRemoteURL)
+		}
+	}
+	if !util.IsInCluster(opt.Karmada.Spec.HostCluster) && opt.Karmada.Spec.Components.KarmadaAPIServer.ServiceType != corev1.ServiceTypeNodePort {
```
Comment (jwcesign) [inline operator/pkg/init.go:52]: If in the public cloud, it could be type LoadBlancer.

File: operator/pkg/util/kubeconfig.go [hunk 3]
```diff
@@ -38,3 +44,40 @@ func CreateBasic(serverURL, clusterName, userName string, caCert []byte) *client
 		CurrentContext: contextName,
 	}
 }
+
+func IsInCluster(hostCluster *operatorv1alpha1.HostCluster) bool {
```
Comment (jwcesign) [inline operator/pkg/util/kubeconfig.go:48]: Can we add some comments about this? To explain how to decide whether it's an in-cluster installation.

File: operator/pkg/init.go [hunk 4]
```diff
@@ -36,6 +37,37 @@ type InitOptions struct {
...
+	}
+	_, err := utilversion.ParseGeneric(opt.KarmadaVersion)
+	if err != nil {
+		return fmt.Errorf("unexpected karmada invalid version %s", opt.KarmadaVersion)
+	}
+
+	if opt.Karmada.Spec.Components.Etcd.Local != nil && opt.Karmada.Spec.Components.Etcd.Local.CommonSettings.Replicas != nil {
+		replicas := *opt.Karmada.Spec.Components.Etcd.Local.CommonSettings.Replicas
+
+		if (replicas % 2) == 0 {
```
Comment (jwcesign) [inline operator/pkg/init.go:63]: Is it a hard limitation?

File: operator/pkg/init.go (hunk 2 above)
Comment (calvin0327) [reply operator/pkg/init.go:52]: Yes, currently, we do not support service of type LB. We haven't figured out how to support the public cloud scenario yet. but this is a good question.

Comment (jwcesign) [reply operator/pkg/init.go:52]: In the public cloud, just set the type as LB, and the public cloud LB controller will handle it.

File: operator/pkg/init.go (hunk 4 above)
Comment (calvin0327) [reply operator/pkg/init.go:63]: No, this is only a warning logs.

File: operator/pkg/init.go (hunk 2 above)
Comment (calvin0327) [reply operator/pkg/init.go]: Ok, I got it and updated.

Comment (calvin0327) [reply operator/pkg/init.go]: @jwcesign PTAL

File: operator/pkg/controller/karmada/planner.go [hunk 5]
```diff
@@ -124,18 +129,44 @@ func (p *Planner) runJobErr(err error) error {
 
 func (p *Planner) afterRunJob() error {
 	if p.action == InitAction {
-		// Update the condition to Ready and set kubeconfig of karmada-apiserver to status.
+		// Update the karmada condition to Ready and set kubeconfig of karmad apiserver to karmada status.
```
Comment (jwcesign) [inline operator/pkg/controller/karmada/planner.go:132]:
```diff
-// Update the karmada condition to Ready and set kubeconfig of karmad apiserver to karmada status.
+// Update the karmada condition to Ready and set kubeconfig of karmada apiserver to karmada status.
```

File: operator/pkg/init.go (hunk 2 above)
Comment (jwcesign) [reply operator/pkg/init.go]: Nice

**Comment by jwcesign** [comment]
/lgtm

cc @RainbowMango for checking

File: operator/pkg/controller/karmada/planner.go (hunk 5 above)
Comment (RainbowMango) [reply operator/pkg/controller/karmada/planner.go:132]: +1
A typo here  "karmad`a`"

Comment (calvin0327) [reply operator/pkg/controller/karmada/planner.go]: @jwcesign @RainbowMango Thanks, updated.

**Review by jwcesign — COMMENTED** [review]
/lgtm

**Review by RainbowMango — APPROVED** [review]
/approve

**Pull request merged**